src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -2317,6 +2319,55 @@ namespace smt::noodler {
         add_axiom({mk_literal(e), ~mk_literal(s_in_digit)});
     }
 
+    /**
+     * @brief Add special axioms for conversions.
+     * In particular, for n = to_int(x) generate n = to_int(x) -> x \in 0*to_string(n)).
+     */
+    void theory_str_noodler::add_conversion_num_axioms() {
+        unsigned nFormulas = ctx.get_num_asserted_formulas();
+        for (unsigned i = 0; i < nFormulas; ++i) {
+            expr *ex = ctx.get_asserted_formula(i);
+            rational val;
+            expr* to_int_arg = nullptr;
+            if(expr_cases::is_to_int_num_eq(ex, m, m_util_s, m_util_a, to_int_arg, val) && val.is_nonneg()) {
+                expr_ref re(m_util_s.re.mk_concat(m_util_s.re.mk_star(m_util_s.re.mk_to_re(m_util_s.str.mk_string("0"))), m_util_s.re.mk_to_re(m_util_s.str.mk_string(val.to_string()))), m);
+                expr_ref in_re(m_util_s.re.mk_in_re(to_int_arg, re), m);
+                add_axiom({~mk_literal(ex), mk_literal(in_re)});
+            }
+        }
+    }
+
+    /**
+     * @brief Add special axioms for length (in)equations. In particular
+     * - for (len s) == 10 create (len s) == 10 -> s \in \Sigma^10
+     * - for (len s) <= 10 create (len s) <= 10 -> s \in re.loop(0, 10)
+     */
+    void theory_str_noodler::add_len_num_axioms() {
+        unsigned nFormulas = ctx.get_num_asserted_formulas();
+        for (unsigned i = 0; i < nFormulas; ++i) {
+            expr *ex = ctx.get_asserted_formula(i);
+            rational val;
+            expr* len_arg = nullptr;
+            if(expr_cases::is_len_num_eq(ex, m, m_util_s, m_util_a, len_arg, val) && val.is_nonneg() && val < 64 && val > 0) {
```

Comment:
is 64 magical constant? maybe define it at the beginning of the function, with some reasoning why 64.

Reply:
I added comments.